publishers/pipeline.go

```diff
@@ -25,26 +24,38 @@ func NewPipeline() *Pipeline {
 	return pipeline
 }
 
-func (pipeline *Pipeline) Publish(packageVersion data.PackageVersion) {
-	pipeline.queue <- packageVersion
+func (pipeline *Pipeline) Publish(ttl time.Duration, packageVersion data.PackageVersion) {
+	pipeline.queue <- publishing{PackageVersion: packageVersion, ttl: ttl}
 }
 
 func (pipeline *Pipeline) run() {
-	pipeline.queue = make(chan data.PackageVersion, maxQueueSize)
+	pipeline.queue = make(chan publishing, maxQueueSize)
 
-	for packageVersion := range pipeline.queue {
-		pipeline.process(packageVersion)
+	for publishing := range pipeline.queue {
+		pipeline.process(publishing)
 	}
 }
 
-func (pipeline *Pipeline) process(packageVersion data.PackageVersion) {
-	// TODO move deduping code here
-	for _, publisher := range pipeline.publishers {
-		// Publish each packageversion to all publishers
-		publisher.Publish(packageVersion)
+func (pipeline *Pipeline) process(publishing publishing) {
+	if pipeline.shouldPublish(publishing) {
```

Comment:
I donno if go would generally rather a 
```
return if !pipeline.shouldPublish(publishing)
```

Reply:
I think I prefer that personally -- seems like Go is consistently nice when it does simple returns like that instead of nested conditionals.